src/marqo/core/distributed_lock/deployment_lock.py

```diff
@@ -0,0 +1,116 @@
+            path: The lock path.
+            max_lock_period: The maximum period of time the lock can be held.
+            watchdog_interval: The interval at which the watchdog checks the lock status.
+            acquire_timeout: The timeout to acquire the lock.
+            error_message: The error message to display when the lock is not acquired.
+        """
+        super().__init__(zookeeper_client, path, max_lock_period, watchdog_interval)
+        self.acquire_timeout = acquire_timeout
+        self.error_message = error_message
+
+    def _watchdog(self):
+        """Watchdog thread to monitor lock status and release the lock if it is held for too long."""
+        while True:
+            if self.lock_acquired_time and (time.time() - self.lock_acquired_time) > self.max_lock_period:
+                self.release()
+                logger.warning(f"Lock released by expiration timer. {self.error_message}")
+            time.sleep(self.watchdog_interval)
+
+    def acquire(self, acquire_timeout: Optional[float] = None) -> bool:
+        """
+        Acquire the lock.
+        Args:
+            acquire_timeout: The timeout to acquire the lock. If not provided, we use the class object's timeout.
+
+        Returns:
+            bool: True if the lock is acquired, False otherwise.
+        """
+        try:
+            acquired = self.lock.acquire(timeout=self.acquire_timeout if acquire_timeout is None else acquire_timeout)
+        except LockTimeout:
+            return False
+        if acquired:
+            self.lock_acquired_time = time.time()
+        return acquired
+
+    def release(self):
+        """Release the lock and reset the lock acquired time."""
+        if self.is_acquired:
+            self.lock.release()
+        self.lock_acquired_time = None
+
+    @property
+    def is_acquired(self) -> bool:
+        return self.lock.is_acquired
+
+
+@contextmanager
+def acquire_deployment_lock(lock: Optional[DeploymentLock] = None, acquire_timeout: Optional[float] = None,
+                            conflict_error_message: Optional[str] = None):
+    """Acquire the deployment lock using a context manager.
+
+    Args:
+        lock: The deployment lock object.
+        acquire_timeout: The timeout to acquire the lock. If not provided, we use the class object's timeout.
+        conflict_error_message: The error message to display when the lock is not acquired. If not provided, we use the
+            class object's error message.
+
+    Raises:
+        RuntimeError: If the lock is not instantiated and called.
+        BackendCommunicationError: If the Zookeeper client cannot connect to the server.
+        ConflictError: If the lock is not acquired.
+    """
+    if lock is None:
+        raise RuntimeError("Deployment lock is not instantiated and cannot be used for index creation/deletion")
+    elif lock.zookeeper_client.state != KazooState.CONNECTED:
+        try:
+            lock.zookeeper_client.start(5)
```

Comment:
Is there a reason not to have the logic to handle Zookeeper connection and errors in your `acquire` method? Accessing `lock.zooekeepr_client` seems to violate the lock class's encapsulation